### PR TITLE
[Data] Refactor sort key

### PR DESCRIPTION
**Pull request by kukushking**
## Why are these changes needed?

As discussed in #37124 with @raulchen,  sort key definition has been inconsistent across the codebase. This change introduces `SortKey` class to wrap sort key and ordering and encapsulates logic to convert between different sort key arguments e.g. pyarrow-style `[(column, "descending")]` or pandas-style `column, True`.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/arrow_block.py [hunk 1]
```diff
@@ -412,9 +412,10 @@ def sum_of_squared_diffs_from_mean(
         )
 
     def sort_and_partition(
-        self, boundaries: List[T], key: "SortKeyT", descending: bool
+        self, boundaries: List[T], sort_key: "SortKey"
     ) -> List["Block"]:
-        if len(key) > 1:
+        columns, ascending = sort_key.to_pandas_sort_args()
```
Comment (raulchen) [inline python/ray/data/_internal/arrow_block.py:417]: nit, this function is not related to pandas, so it's a bit confusing to call `to_pandas_sort_args`. I think we can call 2 separate APIs to get them.

File: python/ray/data/_internal/arrow_ops/transform_polars.py [hunk 2]
```diff
@@ -23,18 +23,18 @@ def check_polars_installed():
...
+    return df.sort(columns, reverse=not ascending).to_arrow()
 
 
 def concat_and_sort(
-    blocks: List["pyarrow.Table"], key: "SortKeyT", descending: bool
+    blocks: List["pyarrow.Table"], sort_key: "SortKey"
 ) -> "pyarrow.Table":
     check_polars_installed()
-    col, _ = key[0]
+    columns, ascending = sort_key.to_pandas_sort_args()
```
Comment (raulchen) [inline python/ray/data/_internal/arrow_ops/transform_polars.py:37]: ditto

File: python/ray/data/block.py [hunk 3]
```diff
@@ -53,7 +53,7 @@
 
 def _validate_key_fn(
```
Comment (raulchen) [inline python/ray/data/block.py:54]: Also move this function to `SortKey`? It can be named as `validate_schema`. and we only  need to handle the `list[str]` case.

**Review by raulchen — APPROVED** [review]
Thanks for your contribution. The PR looks pretty good. I just left a few minor comments.

File: python/ray/data/_internal/arrow_block.py (hunk 1 above)
Comment (kukushking) [reply python/ray/data/_internal/arrow_block.py]: Thanks, added a separate `get_descending()` function, in addition to the existing `get_columns()`.

File: python/ray/data/block.py (hunk 3 above)
Comment (kukushking) [reply python/ray/data/block.py]: Yep good idea - done.

File: python/ray/data/_internal/arrow_block.py [hunk 4]
```diff
@@ -452,7 +447,7 @@ def sort_and_partition(
         partitions.append(table.slice(last_idx))
         return partitions
 
-    def combine(self, key: str, aggs: Tuple[AggregateFn]) -> Block:
+    def combine(self, key: str, aggs: Tuple["AggregateFn"]) -> Block:
```
Comment (kukushking) [inline python/ray/data/_internal/arrow_block.py:450]: Avoids circular dependency singe aggregate class uses` SortKey` for key validation now.
